test/e2e/upgrade/upgrade_test.go

```diff
@@ -178,201 +137,78 @@ func checkRoutability(host string) {
+	require.NoError(f.T(), f.Deployment.EnsureEnvoyService())
 
 	By("updating contour deployment")
-	resources.contourDeployment = new(apps_v1.Deployment)
-	require.NoError(f.T(), decoder.Decode(resources.contourDeployment))
-	tempDep := new(apps_v1.Deployment)
-	require.NoError(f.T(), f.Client.Get(context.TODO(), client.ObjectKeyFromObject(resources.contourDeployment), tempDep))
-	resources.contourDeployment.SetResourceVersion(tempDep.GetResourceVersion())
 	// Update container image.
-	require.Len(f.T(), resources.contourDeployment.Spec.Template.Spec.Containers, 1)
-	resources.contourDeployment.Spec.Template.Spec.Containers[0].Image = contourUpgradeToImage
-	resources.contourDeployment.Spec.Template.Spec.Containers[0].ImagePullPolicy = v1.PullIfNotPresent
-	require.NoError(f.T(), f.Client.Update(context.TODO(), resources.contourDeployment))
+	require.Len(f.T(), f.Deployment.ContourDeployment.Spec.Template.Spec.Containers, 1)
+	f.Deployment.ContourDeployment.Spec.Template.Spec.Containers[0].Image = contourUpgradeToImage
+	f.Deployment.ContourDeployment.Spec.Template.Spec.Containers[0].ImagePullPolicy = v1.PullIfNotPresent
+	require.NoError(f.T(), f.Deployment.EnsureContourDeployment())
 
 	By("updating envoy daemonset")
-	resources.envoyDaemonSet = new(apps_v1.DaemonSet)
-	require.NoError(f.T(), decoder.Decode(resources.envoyDaemonSet))
-	tempDS := new(apps_v1.DaemonSet)
-	require.NoError(f.T(), f.Client.Get(context.TODO(), client.ObjectKeyFromObject(resources.envoyDaemonSet), tempDS))
-	resources.envoyDaemonSet.SetResourceVersion(tempDS.GetResourceVersion())
 	// Update container image.
-	require.Len(f.T(), resources.envoyDaemonSet.Spec.Template.Spec.InitContainers, 1)
-	resources.envoyDaemonSet.Spec.Template.Spec.InitContainers[0].Image = contourUpgradeToImage
-	resources.envoyDaemonSet.Spec.Template.Spec.InitContainers[0].ImagePullPolicy = v1.PullIfNotPresent
-	require.Len(f.T(), resources.envoyDaemonSet.Spec.Template.Spec.Containers, 2)
-	resources.envoyDaemonSet.Spec.Template.Spec.Containers[0].Image = contourUpgradeToImage
-	resources.envoyDaemonSet.Spec.Template.Spec.Containers[0].ImagePullPolicy = v1.PullIfNotPresent
-	require.NoError(f.T(), f.Client.Update(context.TODO(), resources.envoyDaemonSet))
-
-	return resources
+	require.Len(f.T(), f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers, 1)
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers[0].Image = contourUpgradeToImage
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers[0].ImagePullPolicy = v1.PullIfNotPresent
+	require.Len(f.T(), f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers, 2)
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers[0].Image = contourUpgradeToImage
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers[0].ImagePullPolicy = v1.PullIfNotPresent
+	require.NoError(f.T(), f.Deployment.EnsureEnvoyDaemonSet())
+	// TODO: when we deploy the from version in this test, ensure envoy
```

Comment:
Does this need an issue to track?

Reply:
I'll remove TODOs and put them in https://github.com/projectcontour/contour/issues/3679 instead